src/libraries/Common/src/System/Net/Mail/MailAddressParser.cs

```diff
@@ -104,41 +125,73 @@ private static MailAddress ParseAddress(string data, bool expectMultipleAddresse
                 {
                     index--; // Skip the angle bracket
                     // Skip whitespace, but leave comments, as they may be part of the display name.
-                    index = WhitespaceReader.ReadFwsReverse(data, index);
+                    if (!WhitespaceReader.TryReadFwsReverse(data, index, out index, throwExceptionIfFail))
+                    {
+                        Debug.Assert(!throwExceptionIfFail);
+                        return false;
+                    }
                 }
                 else
                 {
-                    // Mismatched angle brackets, throw
-                    throw new FormatException(SR.Format(SR.MailHeaderFieldInvalidCharacter,
-                        (index >= 0 ? data[index] : MailBnfHelper.EndAngleBracket)));
+                    if (throwExceptionIfFail)
+                    {
+                        // Mismatched angle brackets, throw
+                        throw new FormatException(SR.Format(SR.MailHeaderFieldInvalidCharacter,
+                            (index >= 0 ? data[index] : MailBnfHelper.EndAngleBracket)));
+                    }
+                    else
+                    {
+                        return false;
+                    }
                 }
             }
 
             // Is there anything left to parse?
             // There could still be a display name or another address
             if (index >= 0 && !(expectMultipleAddresses && data[index] == MailBnfHelper.Comma))
             {
-                displayName = ParseDisplayName(data, ref index, expectMultipleAddresses);
+                if (!TryParseDisplayName(data, ref index, expectMultipleAddresses, out displayName, throwExceptionIfFail))
+                {
+                    Debug.Assert(!throwExceptionIfFail);
+                    return false;
+                }
             }
             else
             {
                 displayName = string.Empty;
             }
 
-            return new MailAddress(displayName, localPart, domain);
+            parseAddressInfo = new ParseAddressInfo(displayName, localPart, domain);
+            return true;
         }
 
         // Read through a section of CFWS.  If we reach the end of the data string then throw because not enough of the
         // MailAddress components were found.
-        private static int ReadCfwsAndThrowIfIncomplete(string data, int index)
+        private static bool TryReadCfwsAndThrowIfIncomplete(string data, int index, out int outIndex, bool throwExceptionIfFail)
         {
-            index = WhitespaceReader.ReadCfwsReverse(data, index);
+            outIndex = default;
+
+            if (!WhitespaceReader.TryReadCfwsReverse(data, index, out index, throwExceptionIfFail))
+            {
+                Debug.Assert(!throwExceptionIfFail);
```

Comment:
Do we need all of these `Debug.Assert(!throwExceptionIfFail)`s?  I understand the purpose, but it seems like we need to rely on these calls to do the right thing, just like our callers rely on us to do the right thing.

Reply:
Understood your point, I'll move assertion at the end of called method.

Reply:
I'll remove the assertions, I did a check and on called method doesn't make a lot of sense, I should add an assertion on a bool check on same variable.